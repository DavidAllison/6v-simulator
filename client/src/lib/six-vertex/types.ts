/**
 * Core type definitions for the 6-vertex model simulator
 * Based on the square ice model with vertex configurations
 */

/**
 * Vertex types based on the 6 allowed ice configurations
 * Named according to the paper conventions:
 * - a1, a2: Source/sink configurations (2 in, 2 out or vice versa)
 * - b1, b2: Straight-through configurations (horizontal or vertical flow)
 * - c1, c2: Turn configurations (flow changes direction)
 */
export const VertexType = {
  a1: 'a1', // In: left, top    | Out: right, bottom
  a2: 'a2', // In: right, bottom | Out: left, top
  b1: 'b1', // In: left, right   | Out: top, bottom
  b2: 'b2', // In: top, bottom   | Out: left, right
  c1: 'c1', // In: left, bottom  | Out: right, top
  c2: 'c2', // In: right, top    | Out: left, bottom
} as const;

export type VertexType = (typeof VertexType)[keyof typeof VertexType];

/**
 * Edge direction relative to a vertex
 */
export const EdgeDirection = {
  Left: 'left',
  Right: 'right',
  Top: 'top',
  Bottom: 'bottom',
} as const;

export type EdgeDirection = (typeof EdgeDirection)[keyof typeof EdgeDirection];

/**
 * Edge state (arrow direction on the edge)
 * For horizontal edges: In = left-to-right, Out = right-to-left
 * For vertical edges: In = top-to-bottom, Out = bottom-to-top
 */
export const EdgeState = {
  In: 'in',
  Out: 'out',
} as const;

export type EdgeState = (typeof EdgeState)[keyof typeof EdgeState];

/**
 * Represents a position in the lattice
 */
export interface Position {
  row: number;
  col: number;
}

/**
 * Represents an edge in the lattice
 */
export interface Edge {
  from: Position;
  to: Position;
  state: EdgeState;
}

/**
 * Configuration of edges around a single vertex
 */
export interface VertexConfiguration {
  left: EdgeState;
  right: EdgeState;
  top: EdgeState;
  bottom: EdgeState;
}

/**
 * A vertex in the lattice with its type and position
 */
export interface Vertex {
  position: Position;
  type: VertexType;
  configuration: VertexConfiguration;
}

/**
 * The complete lattice state
 */
export interface LatticeState {
  width: number;
  height: number;
  vertices: Vertex[][];
  // Store edge states for efficient access
  horizontalEdges: EdgeState[][]; // [row][col] represents edge to the right of vertex
  verticalEdges: EdgeState[][]; // [row][col] represents edge below vertex
}

/**
 * Boundary conditions for the lattice
 */
export const BoundaryCondition = {
  Periodic: 'periodic', // Toroidal topology
  Fixed: 'fixed', // Fixed boundary arrows
  Open: 'open', // Open boundaries
  DWBC: 'dwbc', // Domain Wall Boundary Conditions
} as const;

export type BoundaryCondition = (typeof BoundaryCondition)[keyof typeof BoundaryCondition];

/**
 * Configuration for DWBC (Domain Wall Boundary Conditions)
 */
export interface DWBCConfig {
  type: 'high' | 'low';
  // High: all top/right edges point in, all bottom/left edges point out
  // Low: opposite of high
}

/**
 * A flip operation that can be performed on the lattice
 */
export interface FlipOperation {
  vertices: Position[]; // Vertices involved in the flip (usually 2)
  edgesToFlip: Edge[]; // Edges that will change state
  isValid: boolean; // Whether this flip maintains ice rule
}

/**
 * Parameters for the Monte Carlo simulation
 */
export interface SimulationParams {
  temperature: number; // Temperature parameter for Metropolis algorithm
  beta: number; // Inverse temperature (1/kT)
  weights: {
    // Boltzmann weights for each vertex type
    a1: number;
    a2: number;
    b1: number;
    b2: number;
    c1: number;
    c2: number;
  };
  boundaryCondition: BoundaryCondition;
  dwbcConfig?: DWBCConfig;
  seed?: number; // Random seed for reproducibility
}

/**
 * Statistics collected during simulation
 */
export interface SimulationStats {
  step: number;
  energy: number;
  vertexCounts: Record<VertexType, number>;
  acceptanceRate: number;
  flipAttempts: number;
  successfulFlips: number;
  beta: number;
  height?: number;
  volume?: number;
  delta?: number;
  entropy?: number;
}

/**
 * Rendering modes for visualization
 */
export const RenderMode = {
  Paths: 'paths', // Show bold edges forming paths
  Arrows: 'arrows', // Show arrow directions on edges
  Both: 'both', // Show both paths and arrows
  Vertices: 'vertices', // Color-code vertex types
} as const;

export type RenderMode = (typeof RenderMode)[keyof typeof RenderMode];

/**
 * Rendering configuration
 */
export interface RenderConfig {
  mode: RenderMode;
  cellSize: number; // Size of each lattice cell in pixels
  lineWidth: number; // Width of edges
  colors: {
    background: string;
    grid: string;
    pathSegment: string; // Bold edges in path mode
    arrow: string;
    vertexTypes: Record<VertexType, string>;
  };
  showGrid: boolean;
  animateFlips: boolean;
  animationDuration: number; // ms
}

/**
 * Observable events emitted during simulation
 */
export interface SimulationEvents {
  onStep: (stats: SimulationStats) => void;
  onFlip: (flip: FlipOperation) => void;
  onStateChange: (state: LatticeState) => void;
  onError: (error: Error) => void;
}

/**
 * Main simulation controller interface
 */
export interface SimulationController {
  // State management
  initialize(width: number, height: number, params: SimulationParams): void;
  reset(): void;
  getState(): LatticeState;
  setState(state: LatticeState): void;

  // Simulation control
  step(): void;
  run(steps: number): void;
  pause(): void;
  resume(): void;
  isRunning(): boolean;

  // Configuration
  updateParams(params: Partial<SimulationParams>): void;
  getParams(): SimulationParams;
  getStats(): SimulationStats;

  // Event handling
  on<K extends keyof SimulationEvents>(event: K, handler: SimulationEvents[K]): void;
  off<K extends keyof SimulationEvents>(event: K, handler: SimulationEvents[K]): void;

  // Data export/import
  exportData(): {
    params: SimulationParams;
    stats: SimulationStats;
    state: LatticeState;
  };
  importData(data: { params: SimulationParams; stats: SimulationStats; state: LatticeState }): void;
}

/**
 * Type guard to check if a configuration represents a valid vertex type
 */
export function getVertexType(config: VertexConfiguration): VertexType | null {
  const { left, right, top, bottom } = config;

  // Count ins and outs
  const ins = [left, right, top, bottom].filter((e) => e === EdgeState.In).length;
  const outs = [left, right, top, bottom].filter((e) => e === EdgeState.Out).length;

  // Ice rule: must have 2 ins and 2 outs
  if (ins !== 2 || outs !== 2) {
    return null;
  }

  // Determine vertex type based on configuration
  if (left === EdgeState.In && top === EdgeState.In) return VertexType.a1;
  if (right === EdgeState.In && bottom === EdgeState.In) return VertexType.a2;
  if (left === EdgeState.In && right === EdgeState.In) return VertexType.b1;
  if (top === EdgeState.In && bottom === EdgeState.In) return VertexType.b2;
  if (left === EdgeState.In && bottom === EdgeState.In) return VertexType.c1;
  if (right === EdgeState.In && top === EdgeState.In) return VertexType.c2;

  return null;
}

/**
 * Get vertex configuration from vertex type
 */
export function getVertexConfiguration(type: VertexType): VertexConfiguration {
  switch (type) {
    case VertexType.a1:
      return {
        left: EdgeState.In,
        top: EdgeState.In,
        right: EdgeState.Out,
        bottom: EdgeState.Out,
      };
    case VertexType.a2:
      return {
        right: EdgeState.In,
        bottom: EdgeState.In,
        left: EdgeState.Out,
        top: EdgeState.Out,
      };
    case VertexType.b1:
      return {
        left: EdgeState.In,
        right: EdgeState.In,
        top: EdgeState.Out,
        bottom: EdgeState.Out,
      };
    case VertexType.b2:
      return {
        top: EdgeState.In,
        bottom: EdgeState.In,
        left: EdgeState.Out,
        right: EdgeState.Out,
      };
    case VertexType.c1:
      return {
        left: EdgeState.In,
        bottom: EdgeState.In,
        right: EdgeState.Out,
        top: EdgeState.Out,
      };
    case VertexType.c2:
      return {
        right: EdgeState.In,
        top: EdgeState.In,
        left: EdgeState.Out,
        bottom: EdgeState.Out,
      };
  }
}
