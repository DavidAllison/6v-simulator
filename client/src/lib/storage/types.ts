/**
 * Type definitions for the storage system
 */

import type { LatticeState, SimulationParams, SimulationStats } from '../six-vertex/types';

/**
 * Saved simulation state that can be persisted
 */
export interface SavedSimulation {
  id: string;
  name: string;
  description?: string;
  createdAt: Date;
  updatedAt: Date;
  version: string; // Version of the save format
  data: any; // Can be compressed string or SimulationData object
  metadata: SimulationMetadata;
}

/**
 * Core simulation data that needs to be saved
 */
export interface SimulationData {
  latticeState: LatticeState;
  params: SimulationParams;
  stats: SimulationStats;
}

/**
 * Additional metadata about the simulation
 */
export interface SimulationMetadata {
  latticeSize: {
    width: number;
    height: number;
  };
  totalSteps: number;
  simulationTime?: number; // Time spent simulating in ms
  compression?: 'none' | 'lz-string'; // Compression method used
  checksum?: string; // Optional checksum for data integrity
}

/**
 * Options for saving a simulation
 */
export interface SaveOptions {
  name: string;
  description?: string;
  compress?: boolean;
  overwrite?: boolean; // Whether to overwrite existing save with same name
}

/**
 * Options for loading a simulation
 */
export interface LoadOptions {
  decompress?: boolean;
  validateChecksum?: boolean;
}

/**
 * Result of a save operation
 */
export interface SaveResult {
  success: boolean;
  id?: string;
  error?: string;
}

/**
 * Result of a load operation
 */
export interface LoadResult {
  success: boolean;
  data?: SimulationData;
  error?: string;
}

/**
 * Storage statistics
 */
export interface StorageStats {
  totalSaves: number;
  totalSize: number; // Total size in bytes
  storageType: 'indexeddb' | 'localStorage';
  maxStorage?: number; // Maximum storage available
}

/**
 * Interface for storage implementations
 */
export interface StorageProvider {
  isAvailable(): Promise<boolean>;
  save(data: SavedSimulation): Promise<SaveResult>;
  load(id: string): Promise<SavedSimulation | null>;
  loadAll(): Promise<SavedSimulation[]>;
  delete(id: string): Promise<boolean>;
  deleteAll(): Promise<boolean>;
  getStats(): Promise<StorageStats>;
}

/**
 * Events emitted by the storage system
 */
export interface StorageEvents {
  onSaveComplete: (result: SaveResult) => void;
  onLoadComplete: (result: LoadResult) => void;
  onDeleteComplete: (id: string) => void;
  onStorageError: (error: Error) => void;
}
