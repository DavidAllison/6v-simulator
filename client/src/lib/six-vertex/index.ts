/**
 * Main export file for the 6-vertex model physics engine
 */

// Core types
export * from './types';

// Vertex shapes and path rendering
export * from './vertexShapes';

// Initial state generators
export {
  generateDWBCHigh,
  generateDWBCLow,
  generateDWBCState,
  generateRandomIceState,
  generateUniformState,
  validateIceRule,
} from './initialStates';

// Physics-accurate flip operations
export {
  FlipDirection,
  isFlippable,
  executeFlip,
  getWeightRatio,
  getAllFlippablePositions,
  calculateHeight,
} from './physicsFlips';
export type { FlipCapability } from './physicsFlips';

// Physics simulation
export { PhysicsSimulation } from './physicsSimulation';
export type { PhysicsSimConfig } from './physicsSimulation';

// Random number generation
export { SeededRNG, globalRNG, createRNG, generateSeed } from './rng';

// Path renderer
export * from './renderer/pathRenderer';

// Test utilities (for development)
export { runTests, validateDWBCHigh, validateDWBCLow, printLattice } from './test';
