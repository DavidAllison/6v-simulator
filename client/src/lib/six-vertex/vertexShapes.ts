/**
 * Maps vertex types to their bold edge patterns for path rendering
 * Based on Figure 1 from David Allison & Reshetikhin (2005) paper: arXiv:cond-mat/0502314v1
 *
 * In the paper's representation:
 * - Bold edges (thick lines) represent paths where arrows point in opposite directions
 * - Thin edges represent arrows pointing in the same direction
 *
 * Vertex type mappings from main.c:
 * - 0 = a1: arrows in from left & top, out to right & bottom
 * - 1 = a2: arrows in from right & bottom, out to left & top
 * - 2 = b1: arrows in from left & right, out to top & bottom
 * - 3 = b2: arrows in from top & bottom, out to left & right
 * - 4 = c1: arrows in from left & bottom, out to right & top
 * - 5 = c2: arrows in from right & top, out to left & bottom
 */

import { VertexType, EdgeDirection } from './types';

/**
 * Path segment definition - which edges connect to form paths through a vertex
 */
export interface PathSegment {
  from: EdgeDirection;
  to: EdgeDirection;
}

/**
 * Get the path segments (bold edges) for a given vertex type
 * Returns pairs of connected edges that form continuous paths
 *
 * Based on Figure 1 of the paper, bold edges occur when arrows on opposite
 * sides of the vertex point in opposite directions (creating a continuous path)
 */
export function getPathSegments(vertexType: VertexType): PathSegment[] {
  // Based on the correct interpretation from when it was working:
  // Bold paths show the ice configuration paths through vertices
  switch (vertexType) {
    case VertexType.a1:
      // Straight paths both horizontal and vertical
      return [
        { from: EdgeDirection.Left, to: EdgeDirection.Right },
        { from: EdgeDirection.Top, to: EdgeDirection.Bottom },
      ];

    case VertexType.a2:
      // Straight paths both horizontal and vertical (reversed)
      return [
        { from: EdgeDirection.Right, to: EdgeDirection.Left },
        { from: EdgeDirection.Bottom, to: EdgeDirection.Top },
      ];

    case VertexType.b1:
      // Vertical straight path only
      return [{ from: EdgeDirection.Top, to: EdgeDirection.Bottom }];

    case VertexType.b2:
      // Horizontal straight path only
      return [{ from: EdgeDirection.Left, to: EdgeDirection.Right }];

    case VertexType.c1:
      // L-shaped turn: left to bottom
      return [{ from: EdgeDirection.Left, to: EdgeDirection.Bottom }];

    case VertexType.c2:
      // L-shaped turn: top to right
      return [{ from: EdgeDirection.Top, to: EdgeDirection.Right }];
  }
}

/**
 * Check if two edges are connected through a vertex (form a path)
 */
export function areEdgesConnected(
  vertexType: VertexType,
  edge1: EdgeDirection,
  edge2: EdgeDirection,
): boolean {
  const segments = getPathSegments(vertexType);
  return segments.some(
    (seg) => (seg.from === edge1 && seg.to === edge2) || (seg.from === edge2 && seg.to === edge1),
  );
}

/**
 * Get the connected edge for a given edge through a vertex
 * Returns null if the edge doesn't connect through
 */
export function getConnectedEdge(
  vertexType: VertexType,
  fromEdge: EdgeDirection,
): EdgeDirection | null {
  const segments = getPathSegments(vertexType);

  for (const segment of segments) {
    if (segment.from === fromEdge) return segment.to;
    if (segment.to === fromEdge) return segment.from;
  }

  return null;
}

/**
 * Visual representation helper for debugging
 * Returns ASCII art representation of vertex type
 */
export function getVertexASCII(vertexType: VertexType): string[] {
  switch (vertexType) {
    case VertexType.a1:
      return ['  ↓  ', '→ ○ →', '  ↓  '];
    case VertexType.a2:
      return ['  ↑  ', '← ○ ←', '  ↑  '];
    case VertexType.b1:
      return ['  ↑  ', '→ ○ →', '  ↓  '];
    case VertexType.b2:
      return ['  ↓  ', '← ○ ←', '  ↑  '];
    case VertexType.c1:
      return ['  ↑  ', '→ ○ ←', '  ↑  '];
    case VertexType.c2:
      return ['  ↓  ', '→ ○ ←', '  ↓  '];
  }
}

/**
 * Get SVG path data for rendering vertex paths
 * Coordinates are relative to vertex center
 */
export function getVertexPathData(
  vertexType: VertexType,
  cellSize: number,
): { paths: string[]; arrows: Array<{ from: [number, number]; to: [number, number] }> } {
  const half = cellSize / 2;
  const paths: string[] = [];
  const arrows: Array<{ from: [number, number]; to: [number, number] }> = [];

  const segments = getPathSegments(vertexType);

  for (const segment of segments) {
    let pathData = '';
    const arrowData: { from: [number, number]; to: [number, number] } | null = null;

    // Create smooth curves for each path segment
    const start = getEdgePoint(segment.from, half);
    const end = getEdgePoint(segment.to, half);

    if (isStrightThrough(segment.from, segment.to)) {
      // Straight line for opposite edges
      pathData = `M ${start[0]} ${start[1]} L ${end[0]} ${end[1]}`;
    } else {
      // Smooth curve for adjacent edges
      pathData = `M ${start[0]} ${start[1]} Q 0 0 ${end[0]} ${end[1]}`;
    }

    paths.push(pathData);

    // Add arrow data for arrow rendering mode
    const arrowStart = getEdgePoint(segment.from, half * 0.5);
    const arrowEnd = getEdgePoint(segment.to, half * 0.5);
    arrows.push({ from: arrowStart, to: arrowEnd });
  }

  return { paths, arrows };
}

/**
 * Get the point on the cell edge for a given direction
 */
function getEdgePoint(direction: EdgeDirection, distance: number): [number, number] {
  switch (direction) {
    case EdgeDirection.Left:
      return [-distance, 0];
    case EdgeDirection.Right:
      return [distance, 0];
    case EdgeDirection.Top:
      return [0, -distance];
    case EdgeDirection.Bottom:
      return [0, distance];
  }
}

/**
 * Check if two edges are opposite (straight through)
 */
function isStrightThrough(edge1: EdgeDirection, edge2: EdgeDirection): boolean {
  return (
    (edge1 === EdgeDirection.Left && edge2 === EdgeDirection.Right) ||
    (edge1 === EdgeDirection.Right && edge2 === EdgeDirection.Left) ||
    (edge1 === EdgeDirection.Top && edge2 === EdgeDirection.Bottom) ||
    (edge1 === EdgeDirection.Bottom && edge2 === EdgeDirection.Top)
  );
}
