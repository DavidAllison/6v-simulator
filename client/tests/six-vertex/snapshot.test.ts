/**
 * Snapshot tests for DWBC patterns and visual regression
 * Verifies exact patterns match paper figures
 */

import {
  generateDWBCHigh,
  generateDWBCLow,
  generateRandomIceState,
  validateIceRule,
} from '../../src/lib/six-vertex/initialStates';
import { VertexType, LatticeState } from '../../src/lib/six-vertex/types';
import { getVertexASCII } from '../../src/lib/six-vertex/vertexShapes';

/**
 * Generate ASCII representation of lattice state for visual verification
 */
function generateLatticeASCII(state: LatticeState): string {
  const lines: string[] = [];
  lines.push(`Lattice ${state.width}x${state.height}:`);

  for (let row = 0; row < state.height; row++) {
    let line = '';
    for (let col = 0; col < state.width; col++) {
      const vertex = state.vertices[row][col];
      // Use simple character representation
      switch (vertex.type) {
        case VertexType.a1:
          line += 'A ';
          break;
        case VertexType.a2:
          line += 'a ';
          break;
        case VertexType.b1:
          line += 'B ';
          break;
        case VertexType.b2:
          line += 'b ';
          break;
        case VertexType.c1:
          line += 'C ';
          break;
        case VertexType.c2:
          line += 'c ';
          break;
        default:
          line += '? ';
          break;
      }
    }
    lines.push(line);
  }

  return lines.join('\n');
}

/**
 * Generate detailed pattern representation
 */
function generateDetailedPattern(state: LatticeState): string[][] {
  const pattern: string[][] = [];

  for (let row = 0; row < state.height; row++) {
    pattern[row] = [];
    for (let col = 0; col < state.width; col++) {
      const vertex = state.vertices[row][col];
      pattern[row][col] = vertex.type;
    }
  }

  return pattern;
}

describe('DWBC Pattern Snapshot Tests', () => {
  describe('DWBC High Pattern (Figure 2)', () => {
    it('should match exact pattern for N=8', () => {
      const state = generateDWBCHigh(8);
      const pattern = generateDetailedPattern(state);

      // Expected pattern for N=8 DWBC High
      // Anti-diagonal should have c2 vertices
      // Upper-left triangle should have b1 vertices
      // Lower-right triangle should have b2 vertices

      const expectedAntiDiagonal = [
        { row: 0, col: 7, type: VertexType.c2 },
        { row: 1, col: 6, type: VertexType.c2 },
        { row: 2, col: 5, type: VertexType.c2 },
        { row: 3, col: 4, type: VertexType.c2 },
        { row: 4, col: 3, type: VertexType.c2 },
        { row: 5, col: 2, type: VertexType.c2 },
        { row: 6, col: 1, type: VertexType.c2 },
        { row: 7, col: 0, type: VertexType.c2 },
      ];

      // Verify anti-diagonal
      for (const { row, col, type } of expectedAntiDiagonal) {
        expect(pattern[row][col]).toBe(type);
      }

      // Verify upper-left triangle (sample check)
      const upperLeftSamples = [
        { row: 0, col: 0, type: VertexType.b1 },
        { row: 0, col: 3, type: VertexType.b1 },
        { row: 2, col: 1, type: VertexType.b1 },
        { row: 3, col: 3, type: VertexType.b1 },
      ];

      for (const { row, col, type } of upperLeftSamples) {
        expect(pattern[row][col]).toBe(type);
      }

      // Verify lower-right triangle (sample check)
      const lowerRightSamples = [
        { row: 7, col: 7, type: VertexType.b2 },
        { row: 7, col: 4, type: VertexType.b2 },
        { row: 5, col: 6, type: VertexType.b2 },
        { row: 4, col: 4, type: VertexType.b2 },
      ];

      for (const { row, col, type } of lowerRightSamples) {
        expect(pattern[row][col]).toBe(type);
      }

      // Store snapshot for visual comparison
      const ascii = generateLatticeASCII(state);
      expect(ascii).toMatchSnapshot('DWBC_High_N8');
    });

    it('should generate correct diagonal pattern for N=24', () => {
      const state = generateDWBCHigh(24);

      // Verify diagonal pattern structure
      let diagonalCorrect = true;
      for (let i = 0; i < 24; i++) {
        if (state.vertices[i][23 - i].type !== VertexType.c2) {
          diagonalCorrect = false;
          break;
        }
      }

      expect(diagonalCorrect).toBe(true);

      // Verify regions
      // Sample upper-left region
      const upperLeftRegion = [];
      for (let row = 0; row < 10; row++) {
        for (let col = 0; col < 10; col++) {
          if (row + col < 10) {
            upperLeftRegion.push(state.vertices[row][col].type);
          }
        }
      }

      const b1Count = upperLeftRegion.filter((t) => t === VertexType.b1).length;
      expect(b1Count).toBe(upperLeftRegion.length);

      // Sample lower-right region
      const lowerRightRegion = [];
      for (let row = 14; row < 24; row++) {
        for (let col = 14; col < 24; col++) {
          if (row + col > 23) {
            lowerRightRegion.push(state.vertices[row][col].type);
          }
        }
      }

      const b2Count = lowerRightRegion.filter((t) => t === VertexType.b2).length;
      expect(b2Count).toBe(lowerRightRegion.length);
    });
  });

  describe('DWBC Low Pattern (Figure 3)', () => {
    it('should match exact pattern for N=8', () => {
      const state = generateDWBCLow(8);
      const pattern = generateDetailedPattern(state);

      // Expected pattern for N=8 DWBC Low
      // Main diagonal should have c2 vertices
      // Upper-right triangle should have a1 vertices
      // Lower-left triangle should have a2 vertices

      const expectedMainDiagonal = [
        { row: 0, col: 0, type: VertexType.c2 },
        { row: 1, col: 1, type: VertexType.c2 },
        { row: 2, col: 2, type: VertexType.c2 },
        { row: 3, col: 3, type: VertexType.c2 },
        { row: 4, col: 4, type: VertexType.c2 },
        { row: 5, col: 5, type: VertexType.c2 },
        { row: 6, col: 6, type: VertexType.c2 },
        { row: 7, col: 7, type: VertexType.c2 },
      ];

      // Verify main diagonal
      for (const { row, col, type } of expectedMainDiagonal) {
        expect(pattern[row][col]).toBe(type);
      }

      // Verify upper-right triangle (sample check)
      const upperRightSamples = [
        { row: 0, col: 1, type: VertexType.a1 },
        { row: 0, col: 7, type: VertexType.a1 },
        { row: 2, col: 5, type: VertexType.a1 },
        { row: 3, col: 4, type: VertexType.a1 },
      ];

      for (const { row, col, type } of upperRightSamples) {
        expect(pattern[row][col]).toBe(type);
      }

      // Verify lower-left triangle (sample check)
      const lowerLeftSamples = [
        { row: 7, col: 0, type: VertexType.a2 },
        { row: 7, col: 3, type: VertexType.a2 },
        { row: 5, col: 1, type: VertexType.a2 },
        { row: 4, col: 3, type: VertexType.a2 },
      ];

      for (const { row, col, type } of lowerLeftSamples) {
        expect(pattern[row][col]).toBe(type);
      }

      // Store snapshot for visual comparison
      const ascii = generateLatticeASCII(state);
      expect(ascii).toMatchSnapshot('DWBC_Low_N8');
    });

    it('should generate correct diagonal pattern for N=24', () => {
      const state = generateDWBCLow(24);

      // Verify diagonal pattern structure
      let diagonalCorrect = true;
      for (let i = 0; i < 24; i++) {
        if (state.vertices[i][i].type !== VertexType.c2) {
          diagonalCorrect = false;
          break;
        }
      }

      expect(diagonalCorrect).toBe(true);

      // Verify regions
      // Sample upper-right region
      const upperRightRegion = [];
      for (let row = 0; row < 10; row++) {
        for (let col = 14; col < 24; col++) {
          if (col > row + 10) {
            upperRightRegion.push(state.vertices[row][col].type);
          }
        }
      }

      const a1Count = upperRightRegion.filter((t) => t === VertexType.a1).length;
      expect(a1Count).toBe(upperRightRegion.length);

      // Sample lower-left region
      const lowerLeftRegion = [];
      for (let row = 14; row < 24; row++) {
        for (let col = 0; col < 10; col++) {
          if (row > col + 10) {
            lowerLeftRegion.push(state.vertices[row][col].type);
          }
        }
      }

      const a2Count = lowerLeftRegion.filter((t) => t === VertexType.a2).length;
      expect(a2Count).toBe(lowerLeftRegion.length);
    });
  });

  describe('Pattern Row/Column Verification', () => {
    it('should have expected vertex types in specific rows for DWBC High', () => {
      const state = generateDWBCHigh(8);

      // Row 0 should have b1 vertices except at column 7 (c2)
      const row0 = state.vertices[0];
      for (let col = 0; col < 7; col++) {
        expect(row0[col].type).toBe(VertexType.b1);
      }
      expect(row0[7].type).toBe(VertexType.c2);

      // Row 7 should have b2 vertices except at column 0 (c2)
      const row7 = state.vertices[7];
      expect(row7[0].type).toBe(VertexType.c2);
      for (let col = 1; col < 8; col++) {
        expect(row7[col].type).toBe(VertexType.b2);
      }

      // Column 0 should have b1 vertices in upper half, c2 at row 7, b2 in lower half
      for (let row = 0; row < 7; row++) {
        expect(state.vertices[row][0].type).toBe(VertexType.b1);
      }
      expect(state.vertices[7][0].type).toBe(VertexType.c2);

      // Column 7 should have c2 at row 0, b2 vertices in rest
      expect(state.vertices[0][7].type).toBe(VertexType.c2);
      for (let row = 1; row < 8; row++) {
        expect(state.vertices[row][7].type).toBe(VertexType.b2);
      }
    });

    it('should have expected vertex types in specific rows for DWBC Low', () => {
      const state = generateDWBCLow(8);

      // Row 0 should have c2 at column 0, then a1 vertices
      expect(state.vertices[0][0].type).toBe(VertexType.c2);
      for (let col = 1; col < 8; col++) {
        expect(state.vertices[0][col].type).toBe(VertexType.a1);
      }

      // Row 7 should have a2 vertices except c2 at column 7
      for (let col = 0; col < 7; col++) {
        expect(state.vertices[7][col].type).toBe(VertexType.a2);
      }
      expect(state.vertices[7][7].type).toBe(VertexType.c2);

      // Column 0 should have c2 at row 0, then a2 vertices
      expect(state.vertices[0][0].type).toBe(VertexType.c2);
      for (let row = 1; row < 8; row++) {
        expect(state.vertices[row][0].type).toBe(VertexType.a2);
      }

      // Column 7 should have a1 vertices except c2 at row 7
      for (let row = 0; row < 7; row++) {
        expect(state.vertices[row][7].type).toBe(VertexType.a1);
      }
      expect(state.vertices[7][7].type).toBe(VertexType.c2);
    });
  });

  describe('Ice Rule Verification in Patterns', () => {
    it('should maintain ice rule in all DWBC patterns', () => {
      const sizes = [4, 6, 8, 12, 16, 24];

      for (const size of sizes) {
        const highState = generateDWBCHigh(size);
        const lowState = generateDWBCLow(size);

        expect(validateIceRule(highState)).toBe(true);
        expect(validateIceRule(lowState)).toBe(true);
      }
    });

    it('should have correct edge configuration at boundaries', () => {
      const state = generateDWBCHigh(6);

      // Check corner vertices satisfy ice rule with boundary
      const topLeft = state.vertices[0][0];
      const topRight = state.vertices[0][5];
      const bottomLeft = state.vertices[5][0];
      const bottomRight = state.vertices[5][5];

      // Each corner should still have 2 in, 2 out counting boundary edges
      expect(topLeft.type).toBe(VertexType.b1);
      expect(topRight.type).toBe(VertexType.c2);
      expect(bottomLeft.type).toBe(VertexType.c2);
      expect(bottomRight.type).toBe(VertexType.b2);
    });
  });

  describe('Pattern Consistency', () => {
    it('should generate identical patterns for same size', () => {
      const state1 = generateDWBCHigh(10);
      const state2 = generateDWBCHigh(10);

      for (let row = 0; row < 10; row++) {
        for (let col = 0; col < 10; col++) {
          expect(state1.vertices[row][col].type).toBe(state2.vertices[row][col].type);
        }
      }
    });

    it('should scale patterns correctly', () => {
      // Smaller patterns should be subsets of larger ones
      const small = generateDWBCHigh(4);
      const large = generateDWBCHigh(8);

      // The pattern structure should be consistent
      // Anti-diagonal in both should have c2
      expect(small.vertices[0][3].type).toBe(VertexType.c2);
      expect(large.vertices[0][7].type).toBe(VertexType.c2);

      expect(small.vertices[3][0].type).toBe(VertexType.c2);
      expect(large.vertices[7][0].type).toBe(VertexType.c2);
    });
  });

  describe('Visual Pattern ASCII Generation', () => {
    it('should generate readable ASCII representation', () => {
      const state = generateDWBCHigh(4);
      const ascii = generateLatticeASCII(state);

      expect(ascii).toContain('Lattice 4x4:');
      expect(ascii.split('\n')).toHaveLength(5); // Title + 4 rows

      // Each row should have 4 vertices represented
      const lines = ascii.split('\n');
      for (let i = 1; i <= 4; i++) {
        const vertices = lines[i].trim().split(/\s+/);
        expect(vertices).toHaveLength(4);
      }
    });

    it('should use consistent notation for vertex types', () => {
      const state = generateDWBCLow(4);
      const ascii = generateLatticeASCII(state);

      // Should contain expected characters
      expect(ascii).toMatch(/[AaBbCc]/);

      // Count vertex representations
      const charCounts: Record<string, number> = {};
      for (const char of ascii.replace(/[^AaBbCc]/g, '')) {
        charCounts[char] = (charCounts[char] || 0) + 1;
      }

      // Total should be 16 (4x4)
      const total = Object.values(charCounts).reduce((a, b) => a + b, 0);
      expect(total).toBe(16);
    });
  });
});
