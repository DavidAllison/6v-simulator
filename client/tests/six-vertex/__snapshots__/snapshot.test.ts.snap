// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`DWBC Pattern Snapshot Tests DWBC High Pattern (Figure 2) should match exact pattern for N=8: DWBC_High_N8 1`] = `
"Lattice 8x8:
B B B B B B B c 
B B B B B B c b 
B B B B B c b b 
B B B B c b b b 
B B B c b b b b 
B B c b b b b b 
B c b b b b b b 
c b b b b b b b "
`;

exports[`DWBC Pattern Snapshot Tests DWBC Low Pattern (Figure 3) should match exact pattern for N=8: DWBC_Low_N8 1`] = `
"Lattice 8x8:
c A A A A A A A 
a c A A A A A A 
a a c A A A A A 
a a a c A A A A 
a a a a c A A A 
a a a a a c A A 
a a a a a a c A 
a a a a a a a c "
`;
